tests/unit/s2n_server_hello_retry_test.c

```diff
@@ -56,6 +106,118 @@ static int client_hello_detect_duplicate_calls(struct s2n_connection *conn, void
     return 0;
 }
 
+int s2n_client_hello_poll_cb(struct s2n_connection *conn, void *ctx)
+{
+    struct client_hello_context *client_hello_ctx;
+    if (ctx == NULL) {
+        return -1;
+    }
+    client_hello_ctx = ctx;
+    /* Increment counter to ensure that callback was invoked */
+    client_hello_ctx->invocations++;
+
+    if (client_hello_ctx->mark_done) {
+        EXPECT_SUCCESS(s2n_client_hello_cb_done(conn));
+        return S2N_SUCCESS;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_negotiate_nonblocking_poll(struct s2n_connection *conn,
+    struct client_hello_context *ch_ctx)
+{
+    s2n_blocked_status blocked;
+    EXPECT_NOT_NULL(conn);
+
+    EXPECT_EQUAL(ch_ctx->invocations, 0);
+
+    /* negotiate handshake, we should pause after the nonblocking callback is invoked */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+    EXPECT_EQUAL(ch_ctx->invocations, 1);
+
+    /* unless explicitly unblocked we should stay paused */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+    EXPECT_EQUAL(ch_ctx->invocations, 2);
+
+    ch_ctx->mark_done = true;
+
+    /* Expect the callback to complete after 2nd iteration */
+    EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
+    EXPECT_EQUAL(ch_ctx->invocations, 3);
+
+    return S2N_SUCCESS;
+}
+
+S2N_RESULT hello_retry_client_hello_cb_test(bool enable_poll) {
+        struct s2n_config *server_config;
+        struct s2n_config *client_config;
+
+        struct s2n_connection *server_conn;
+        struct s2n_connection *client_conn;
+
+        struct s2n_cert_chain_and_key *tls13_chain_and_key;
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(client_config, "default_tls13"));
```

Comment:
Why set up the client security policy / cipher preferences when you're just going to override them later? You should be able to use the same config for both connections.

Reply:
the settings are copied from the test above.